pkg/apis/eventing/v1alpha1/parameter_argument_types.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"github.com/knative/pkg/apis"
+)
+
+// Check that ParameterSpec can be validated and can be defaulted.
+var _ apis.Validatable = (*ParameterSpec)(nil)
+var _ apis.Defaultable = (*ParameterSpec)(nil)
+
+// ParameterSpec allows specification of parameters that optionally have defaults.
+type ParameterSpec struct {
+	// Name is the unique name of this template parameter.
+	// +required
+	Name string `json:"name"`
+
+	// Description is A human-readable explanation of this template parameter.
+	// +optional
+	Description string `json:"description,omitempty"`
+
+	// Required specifies if the parameter is required.
+	// If true, default / defaultFrom should not be provided.
+	// Defaults to false.
+	// +immutable
+	// +optional
+	Required bool `json:"required,omitempty"`
+
+	// Default value if not provided by arguments.
+	// +oneof(Default, DefaultFrom), optional.
+	Default string `json:"default,omitempty"`
+
+	// DefaultFrom specifies a Default value retrieved from an existing Secret
+	// or ConfigMap if not provided by arguments.
+	// +oneof(Default, DefaultFrom), optional.
+	DefaultFrom *ArgumentValueReference `json:"defaultFrom,omitempty"`
```

Comment:
Do we have a use case for this? It smells YAGNI.

Reply:
it was in the spec, so I added it here.

Reply:
Let me go there to make a comment.